File: scm-csharp/scheme/SchemeSocket.cs
using System.Net.Sockets;
using System.IO;

namespace scheme;

public class SchemeSocket
{
    public TcpClient client;
    public TextStream inputPort;
    public TextOutputStream outputPort;
    public Stream networkStream;
    public BinaryInputStream? binaryInputPort;
    public BinaryOutputStream? binaryOutputPort;

    public SchemeSocket(TcpClient client)
    {
        this.client = client;
        this.networkStream = client.GetStream();
        this.inputPort = new TextStream(new StreamReader(networkStream), "{socket}");
        this.outputPort = new TextOutputStream(new StreamWriter(networkStream) { AutoFlush = false });
    }

    public SchemeSocket(TcpClient client, Stream stream)
    {
        this.client = client;
        this.networkStream = stream;
        this.inputPort = new TextStream(new StreamReader(stream), "{socket}");
        this.outputPort = new TextOutputStream(new StreamWriter(stream) { AutoFlush = false });
    }

    public override string ToString() => "#<socket>";
}

File: scm-java/src/main/scheme/SchemeSocket.java
package scheme;

import java.io.*;
import java.net.Socket;

public class SchemeSocket {
    public Socket socket;
    public TextStream inputPort;
    public TextOutputStream outputPort;
    public InputStream networkInputStream;
    public OutputStream networkOutputStream;
    public BinaryInputStream binaryInputPort;
    public BinaryOutputStream binaryOutputPort;

    public SchemeSocket(Socket socket) throws IOException {
        this.socket = socket;
        this.networkInputStream = socket.getInputStream();
        this.networkOutputStream = socket.getOutputStream();
        this.inputPort = new TextStream(
            new java.io.PushbackReader(new BufferedReader(new InputStreamReader(networkInputStream), 8192)),
            "{socket}");
        this.outputPort = new TextOutputStream(new BufferedWriter(new OutputStreamWriter(networkOutputStream), 8192));
    }

    public SchemeSocket(Socket socket, InputStream in, OutputStream out) throws IOException {
        this.socket = socket;
        this.networkInputStream = in;
        this.networkOutputStream = out;
        this.inputPort = new TextStream(
            new java.io.PushbackReader(new BufferedReader(new InputStreamReader(in), 8192)),
            "{socket}");
        this.outputPort = new TextOutputStream(new BufferedWriter(new OutputStreamWriter(out), 8192));
    }

    @Override
    public String toString() { return "#<socket>"; }
}
